feat: user management feature with auth0

vickywane opened the pull request:
This closes [this](https://github.com/ambianic/ambianic-subscriptions/issues/1)

This PR contains work done to implement the new user management feature. 

**Note:** This PR is an active WIP. 

A test of the work being done can be found in the Staging PWA [here](https://staging-ambianic.netlify.app/).

[This](https://drive.google.com/file/d/13ErxNk228X-0gmiNw5Jl4RHzwFwB1P46/view?usp=sharing) design flow diagram visualizes the steps performed by different actors to achieve a user subscription.

![user-subscription (2)](https://user-images.githubusercontent.com/29664439/114277274-6a6c9a80-9a22-11eb-9ee2-822b842fd274.png)

ivelin reviewed (changes requested):
@vickywane good call today.

I see that several checks fail the CI. Can you please clear all errors and also polish the rough edges that I experienced today. Users should not have to manually reload a page to get the expected behavior.

Thank you!

vickywane commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,166 @@
+    },
+    submitUserId () {
+      this.edgeAPI
+        .initializePremiumNotification(this.$auth.user.sub)
```
The code is located [here](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/src/remote/edgeAPI.js#L75). The endpoint is stored in the PWA environment variables which are to be added on Netlify before the PWA is deployed. 

It is part of the sync process which happens between the PWA and the connected Edge device.

vickywane commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,290 @@
+        </v-card-title>
+        <v-card-text data-cy="detail">
+          Subscribe to Ambianic Premium Services to enable advanced
+          features, such as Notification, Multiple Edge Displayed, Reports.
```
Noted. It has been corrected.

vickywane commented on src/auth/index.js:
```diff
@@ -0,0 +1,117 @@
+      loginWithRedirect (o) {
+        return this.auth0Client.loginWithRedirect(o)
+      },
+      handleTestLogin () {
```
It is a mock login function being used by the test runners only. 

It is only accessed by the test runner to render a test authenticated state. For example, the test runner authenticates using the `handleTestLogin` and the UI for authenticated users is shown and tested.  

The user details are stored in the `user` object and are supplied by the `auth0Client`. To test, the `handleTestLogin` method is called by only the test runners within the test suites and non-sensitive user details are supplied to the `user` object.

ivelin replied:
Please use jest mocks to replace the deployment login function at runtime with a mock one for testing. Don't hard code test code in the deployment code. See examples [here](https://github.com/ambianic/ambianic-ui/tree/master/tests/unit/store).

vickywane replied:
This has been implemented.

vickywane commented on package.json:
```diff
@@ -15,11 +15,18 @@
     "konva": "^7.1.4",
     "moment": "^2.26.0",
+    "moment-range": "^4.0.2",
+    "moxios": "^0.4.0",
```
`Moxios` is specially used to `intercept` Axios requests within the test suites for testing. 

If you are strongly in support of the browser `fetch` over `axios`, I would have to look into using jest mock() and remove `moxios`. 

However, it would require some significant refactoring of the tests. Would you prefer it comes in a separate PR?

vickywane replied:
Fetch has now been switched to in the latest commits.

vickywane commented on package.json:
```diff
@@ -15,11 +15,18 @@
+    "@auth0/auth0-spa-js": "^1.14.0",
     "@vue/babel-preset-app": "^4.5.10",
+    "axios": "^0.21.1",
+    "axios-request-handler": "^1.0.4",
```
`Axios` is most times my preferred option, however, I can refactor to use `fetch()`.

Would you prefer it comes in a separate PR?

vickywane replied:
Fetch has now been switched to in the latest commit.

ivelin replied:
OK, then `"axios-request-handler": "^1.0.4"` can be removed from package.json.

vickywane replied:
It has been removed.

vickywane commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,290 @@
+    ...mapActions(['HANDLE_SUBSCRIPTION_DIALOG', 'FETCH_USER_SUBSCRIPTION', 'HANDLE_EDGE_SYNC_DIALOG']),
+    submitSubscription () {
+      this.loading = true
+      Axios.post(
```
I do not think Axios is outdated, neither does the article you attached provide such information. 

However, it compares `axios` with `fetch`.
For our usecase, `axios` appears to be an overkill even though useful.

vickywane commented on src/auth/index.js:
```diff
@@ -0,0 +1,117 @@
+// fix for window.crypto is required bug. See #522
+export const createAuthLink = async () =>
+  !navigator.userAgent.includes('jsdom') &&
+  (await createAuth0Client({
```
For now, i think the major parts are covered in test suites. All new code files introduced have over 60% coverage and I plan to increase the amounts as I keep working on this ( after release ).

ivelin replied:
Any reason why coverage can't be 100% for all new code introduced in this PR?

ivelin commented on package.json:
```diff
@@ -15,12 +15,18 @@
     "idb": "^6.0.0",
     "konva": "^7.1.4",
     "moment": "^2.26.0",
+    "moment-range": "^4.0.2",
```
Where is `moment-range` used in code?

vickywane replied:
It was been used previously to determine if a subscription was active using the `duration` object. This has been changed. I have uninstalled the library.

vickywane commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,166 @@
+          </v-icon>
+
+          <p id="explanation">
+            Your premium subscription has been successfully extended to your connected Edge Device. <br> <br> You would now get email notifications about object detections from your running edge device.
```
Okay. It has been done

vickywane commented on package.json:
```diff
@@ -15,12 +15,18 @@
     "moment": "^2.26.0",
+    "moment-range": "^4.0.2",
     "node": "^15.4.0",
+    "node-fetch": "^2.6.1",
```
I have uninstalled it. I installed it when I was trying to implement the mock requests.

vickywane commented on src/remote/edgeAPI.js:
```diff
@@ -70,4 +70,25 @@ export class EdgeAPI {
+      url: `${apiRoot}auth/premium-notification`,
+      params: {
+        userId,
+        notification_endpoint: process.env.VUE_APP_FUNCTIONS_ENDPOINT
```
I have added this to the readme in  the Environment Variables section.

ivelin commented on README.md:
````diff
@@ -56,6 +56,14 @@ cd ambianic-ui
+To use the notifications feature, the PWA requires the `VUE_APP_FUNCTIONS_ENDPOINT` which should point to the [Cloud API](https://github.com/ambianic/ambianic-subscriptions) either running in production or locally.
+
+```
+VUE_APP_FUNCTIONS_ENDPOINT=<CLOUD_API_ENDPOINT>/.netlify/functions
````
@vickywane it's good to have this doc section. Let's remove VUE from the variable name. It does not matter if we use react or vue, the endpoint for cloud API will be the same.

vickywane replied:
We need to have the `VUE_APP` prefixed to the variable name for value-loaded automatically without using `dotenv`.

ivelin commented on package.json:
```diff
@@ -15,12 +15,18 @@
+    "axios-request-handler": "^1.0.4",
+    "codecov": "^3.7.0",
     "compression-webpack-plugin": "^6.0.4",
+    "flush-promises": "^1.0.2",
```
@vickywane `flush-promises` is still in the deployment `dependencies` section , instead of `devDependencies`

vickywane replied:
`flush-promises` has been removed.

ivelin commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,294 @@
+
+export default {
+  name: 'SubcriptionDialog',
+  data: () => ({
```
@vickywane why is this test code still in the PR? Do we want end users to see this fake card data?

ivelin reviewed (changes requested):
@vickywane a couple comments inline. Good progress.

ivelin reviewed (changes requested):
@vickywane see inline comments.

Also, package-lock.json now has merge conflicts that need to be cleaned up.

vickywane commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,286 @@
+        v-if="subscriptionError"
+      >
+        <p class="error-text">
+          An error occurred while adding your subscription. Please try again
```
Yes. 

There is a plan to include more detailed error logging in the `Settings` page of the Ambianic PWA. This includes the redesigning of the Settings page and is being documented and tracked in this issue, however that issue is dependent on this.

I would improve the error message in the `SubscriptionDialog` component based on your recommendation. 

As discussed on Slack, integrating an external error monitoring tool like [Sentry](https://sentry.io) is being considered, however, it may be a long-term objective.

ivelin replied:
@vickywane  OK, please open an issue to track external cloud API monitoring.

vickywane commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,286 @@
+            data-cy="price"
+            style="font-weight: 500; text-align: center;"
+          >
+            $5 Monthly Fee
```
For now, we can use a hardcoded subscription fee. 

I would look into the Stripe docs on pulling details about a created product. However, I think it might be overkill to make network requests just for that. Any further thoughts?

ivelin replied:
How about cashing the current user subscription fee in the Auth0 user profile. The user profile has to be fetched anyways on login.

vickywane commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,170 @@
+          </v-icon>
+
+          <p id="explanation">
+            Your premium subscription has been successfully extended to your connected Edge Device. <br> <br> You would now get email notifications about detections events from your running edge device.
```
For outdated edge devices, I would add a text to the UI urging the user to ensure the edge device is updated. 

For now, a user has the option to re-sync the edge device when it seems not to work. Advanced error handling still seems to be an advanced use-case and subsequent pull requests would be handling them.

ivelin replied:
This is not advanced. It is a basic UI behavior that needs to be implemented. If the edge device is an outdated version, we need to inform the user about it and advise on corrective actions.

ivelin commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,177 @@
+
+          <hr>
+          <div class="flex" style="margin: .5rem 0 0 0" >
+            <div class="align-center" >
```
@vickywane please resolve all lint errors.

![Screen Shot 2021-06-08 at 9 26 45 AM](https://user-images.githubusercontent.com/2234901/121203410-b037ba80-c83b-11eb-87c7-0df15d90bc84.png)

vickywane replied:
I have fixed them.

ivelin reviewed (changes requested):
@vickywane a few open comments.

vickywane commented:
@ivelin I have updated the premium sync to handle outdated edge devices. You might want to check the texts in the `edge-auth0-sync` component. 

Also, I added a link to the component for users with outdated edge devices to update their devices. The new docs haven't been published so I added a link to the docs.

vickywane commented:
> How about cashing the current user subscription fee in the Auth0 user profile. The user profile has to be fetched anyways on login.

That would make it more difficult to change in the future. Would we have to update each user's profile data because we need to adjust the price. 

Fetching the data from Stripe ( if possible ) would be a better option and I would implement it in a separate PR.

ivelin commented:
> > How about cashing the current user subscription fee in the Auth0 user profile. The user profile has to be fetched anyways on login.
> 
> That would make it more difficult to change in the future. Would we have to update each user's profile data because we need to adjust the price.
> 
> Fetching the data from Stripe ( if possible ) would be a better option and I would implement it in a separate PR.

OK, open an issue to track this discussion.

ivelin commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,231 @@
+            >
+              <div class="align-center">
+                <p style="margin: 0 .3rem; font-size: .95rem">
+                  Please restart your Ambianic Edge device to update your Ambianic Edge Installation.
```
@vickywane restart is not required for ambianic edge to pull an update. docker watchtower does that automatically.
A more appropriate message might be:
"Your Ambianic Edge device is running an outdated software version. It should automatically update to the latest version within 24 hours. Please try again later or contact support if the issue persists."

ivelin commented on src/components/edge-auth0-sync.vue:
```diff
@@ -0,0 +1,231 @@
+              </v-icon>
+
+              <p id="explanation">
+                Your premium subscription has been successfully extended to your connected Edge Device. <br> <br> You would now get email notifications about detections events from your running edge device.
```
@vickywane a small language correction. Please replace:
"You would now get email notifications about detections events from your running edge device." with

"You will  now receive email notifications about detections events from your running edge device."

vickywane replied:
I have made this correction

ivelin reviewed (changes requested):
@vickywane The improved error reporting to end users is nice. See a few comments on it.

Also, we still have an [issue with coverage reporting](https://github.com/ambianic/ambianic-ui/pull/562/checks?check_run_id=2788392394#step:8:55). If you cannot resolve it with codecov support, add a summary text report dump in the CI flow itself so we know if we are regressing.

Also please update the CI script to fail when codecoverage report upload fails. Otherwise it may fly undetected under the radar and allow for regressions.

```
->  Pinging Codecov
https://codecov.io/upload/v4?package=github-action-v1.5.2-1.0.3&token=secret&branch=feature%2Fuser-management&commit=072aaf9b13d0c02fc14ef954b3ac767b435ace94&build=923278782&build_url=http%3A%2F%2Fgithub.com%2Fambianic%2Fambianic-ui%2Factions%2Fruns%2F923278782&name=&tag=&slug=ambianic%2Fambianic-ui&service=github-actions&flags=&pr=562&job=CI%20for%20Ambianic%20UI&cmd_args=n,F,Q,Z,f,C,q
{"error": "Server Error (500)"}{"error": "Server Error (500)"}{"error": "Server Error (500)"}{"error": "Server Error (500)"}{"error": "Server Error (500)"}{"error": "Server Error (500)"}
500
==> Uploading to Codecov
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
```

vickywane commented:
> Also, we still have an issue with coverage reporting. If you cannot resolve it with codecov support, add a summary text report dump in the CI flow itself so we know if we are regressing.

`text-summary` has been added to the reporters being used with Jest. Would that be sufficient?

vickywane commented:
> OK, open an issue to track this discussion.

This [feature request](https://github.com/ambianic/ambianic-ui/issues/652) has been opened to track this issue.

ivelin commented:
> > Also, we still have an issue with coverage reporting. If you cannot resolve it with codecov support, add a summary text report dump in the CI flow itself so we know if we are regressing.
> 
> `text-summary` has been added to the reporters being used with Jest. Would that be sufficient?

@vickywane 

No, that's not enough. The actual report file has to be either dumped in the CI log so we can see the results or the file can be uploaded as a github action artifact so we can open it after the CI run is finished.

ivelin commented:
@vickywane we are getting very close. 

Please let me know the exact environment variables that need to be made available in the netlify deployment in order for the new code to work properly with auth0 and the cloud API.

ivelin reviewed (changes requested):
@vickywane almost there
1. coverage report file contents needs to be either show in the CI log or stored as a github action artifact.
2. Need list of netlify deployment variables.

vickywane commented:
> Need a list of netlify deployment variables.

[This](https://github.com/ambianic/ambianic-subscriptions/pull/19/) pull request was opened to include the full list of environment variables in the Cloud API documentation.

ivelin reviewed (changes requested):
@vickywane good thing code coverage is back online.
It shows a lot of uncovered code. Please take care of that so we can merge.

https://github.com/ambianic/ambianic-ui/pull/562/checks?check_run_id=2830439823

![Screen Shot 2021-06-15 at 1 36 01 PM](https://user-images.githubusercontent.com/2234901/122105708-b64f0d80-cdde-11eb-9eff-88b6c6d05e35.png)

ivelin commented on src/components/NavBar.vue:
```diff
@@ -46,37 +29,45 @@
       />
 
-      <nav-button
+      <amb-button
```
@vickywane why are you adding back the bell icon. It was commented out because it does not do anything meaningful.

vickywane replied:
I think i started working on this feature when the icons were still in use.

ivelin commented on src/components/NavBar.vue:
```diff
@@ -46,37 +29,45 @@
+      />
+
+      <!-- test -->
+      <amb-button
```
@vickywane why are you adding back the heart icon? It was removed in a previous commit, because it does not do anything useful.

vickywane replied:
I think i started working on this feature when the icons were still in use. I would remove them

ivelin replied:
Which is why you need to `pull` from upstream main frequently and be careful with merging so you don't override more recent work.

ivelin commented on src/components/NavBar.vue:
```diff
@@ -177,10 +168,17 @@ import {
   },
   data: () => ({
-    dialog: false,
+    // TESTING POPOVER
```
@vickywane what is the meaning of this code? If its test code, it does not belong in live deployment files?

vickywane replied:
I have removed it.

ivelin commented on src/components/subscriptionDialog.vue:
```diff
@@ -0,0 +1,279 @@
+        }
+      )
+      this.loading = false
+      this.$store.dispatch(HANDLE_SUBSCRIPTION_DIALOG, false)
```
@vickywane did you mean to `await` for these dispatch actions? Otherwise their order of execution is not guaranteed and this `saveStripeData` function will return immediately, before the actions complete their work.

vickywane replied:
The asynchronous execution of some of the action do not matter to the application logic as some of them are background operations.  

Those whose order of execution matter I used the `await` keyword for. I have however made this correcton

ivelin commented on tests/unit/store/premium-service.spec.js:
```diff
@@ -0,0 +1,148 @@
+      id: 'cus|123456789'
+    }
+
+    await store.dispatch('FETCH_USER_SUBSCRIPTION', data)
```
@vickywane where is the condition we want verified at the end of this test?

vickywane replied:
This slipped. I have completed the test suite.

vickywane commented on package.json:
```diff
@@ -12,9 +12,10 @@
     "test:unit": "jest",
     "test:e2e": "npx cypress run",
-    "test": "npm run test:unit && npm run test:e2e"
+    "test": "mkdir -p coverage/jest && npm run test:unit > coverage/jest/coverage.txt && npm run test:e2e"
```
@ivelin 

1. Jest does **not** output a `.txt` file for text coverage, rather it outputs it to the console for the developer to see. It can however output a `.json` file ( if you would prefer this ). This was the reason why I used the `>` hack. See [docs](https://jestjs.io/docs/configuration#coveragereporters-arraystring--string-options)

vickywane replied:
I have implemented 2 & 3. I didn't use the `merge-cypress-jest-coverage` package. I wrote some bash commands to move the file around.

ivelin replied:
@vickywane can you share link to the code that solved 1,2 &3?

ivelin commented on cypress/integration/ambianic-tests/navbar.spec.js:
```diff
@@ -1,23 +1,26 @@
   })
 
+  it('Should have a search bar', () => {
+    cy.get('[data-cy=container').find('#searchbar')
```
@vickywane what post condition do we expect on this test?

ivelin replied:
We should probably disable the search bar until its ready to provide some value.

vickywane replied:
> what post condition do we expect on this test?

I didnt write that test.

> We should probably disable the search bar until its ready to provide some value.

Okay. We can do this.

ivelin commented on cypress/integration/ambianic-tests/navbar.spec.js:
```diff
@@ -1,23 +1,26 @@
   })
 
-
-  /** future buttons
```
@vickywane why did you put back on the buttons that were disabled due to lack of value at this time?

ivelin commented on cypress/integration/ambianic-tests/navbar.spec.js:
```diff
@@ -1,23 +1,26 @@
-      cy.get('[data-cy=container').find("#searchbar")
-  })
-
-  it('Should not have a heart button', () => {
```
@vickywane I don't understand why you keep overriding newer code with outdated one instead of carefully reviewing the merge conflicts. As a best practice you should regularly `pull` from the upstream main branch to minimize accumulation of merge conflicts.

ivelin commented on cypress/integration/ambianic-tests/profile-menu.spec.js:
```diff
@@ -0,0 +1,41 @@
+
+  it('Should open subscription modal', () => {
+    cy.get('[data-cy=add-subscription]').click()
+    cy.get('[data-cy=dismiss-modal]').click()
```
@vickywane it seems like in some tests you are using `click()` as an assertion, but its not intended to be used that way. There has to be a `should` or `expect` in the test case to actually assert some condition. See this blog:
https://www.cypress.io/blog/2019/01/22/when-can-the-test-click/

ivelin commented on cypress/integration/ambianic-tests/profile-menu.spec.js:
```diff
@@ -0,0 +1,41 @@
+    })
+
+    cy.get('#vue-tour-button').click()
+    cy.get('[data-cy=profile-toggle]').click()
```
@vickywane what assertion are we checking in this test? Code coverage is meaningless if there is no thoughtful behavior validation.

ivelin commented on src/auth/index.js:
```diff
@@ -0,0 +1,127 @@
+
+            const user = await this.auth0Client.getUser()
+
+            this.$store.dispatch('SAVE_AUTHENTICATED_USER', {
```
@vickywane similar issue here as before. Because the code does not `await` for async call nor does it check the result of returned promises, there is no easy way of knowing what the execution order is, nor what the outcome is from these calls. Makes it very hard to debug issues.

vickywane replied:
@ivelin [This](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/package.json#L14) is the link to the script that combines the coverage files from jest and cypress. [This](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/package.json#L14) is also the command within the `workflow` file to generate the reports within the workflow build.

It was originally implemented in [this](https://github.com/bahmutov/cypress-and-jest/blob/master/package.json) project.

ivelin commented on src/auth/index.js:
```diff
@@ -0,0 +1,127 @@
+              isAuthenticated: user && true
+            })
+
+            this.$store.dispatch('FETCH_USER_SUBSCRIPTION', user && user.sub)
```
`await`?

vickywane replied:
The `await` keyword has been added to all asynchronous store methods.

ivelin reviewed (changes requested):
Still too many loose ends. I marked a few, but at this point, I am inclined on postponing another review until you are sure that the code has been carefully swept for repetitive types of problems that have been brought up in previous reviews.

vickywane commented:
@ivelin 

Since the last review i have made the following changes:

- I have made improvement on the following e2e test in `subscription.spec.js` file
- I have renamed `profile-menu.spec.js` to `authbar-menu.spec.js` to indicate it's usage
- I have also adjusted the new code to either use a promise when needed or not return any promise rather than ignoring them
- I have updated the `Timeline.vue & Navbar.vue` components to be in sync with the master branch. 
- I have gone through the all new unit tests for this feature and completed them. Some of the unit tests had misleading names and were missing useful assertions.

ivelin commented on package.json:
```diff
@@ -7,14 +7,17 @@
+    "codecov:unit": "codecov -f ./jest-coverage/coverage-final.json",
+    "codecov": "codecov -f jest-coverage/coverage-final.json && codecov -f cypress-coverage/coverage-final.json",
+    "copy:reports": "mkdir -p reports/ && cp cypress-coverage/coverage-final.json reports/from-cypress.json && cp jest-coverage/coverage-final.json reports/from-jest.json",
+    "combine:reports": "mkdir -p reports/ && npx nyc merge reports && mv coverage.json .nyc_output/out.json",
```
@vickywane I see your script to merge cov tests in .nyc_output/out.json, but I don't see where this combined file is used. For example, a txt version of it can be uploaded as the final github action coverage artifact. Currently the CI script only uploads the jest cov report.

ivelin commented on package.json:
```diff
@@ -7,14 +7,17 @@
     "test:unit": "jest",
     "test:e2e": "npx cypress run",
-    "test": "npm run test:unit && npm run test:e2e"
+    "test": "mkdir -p jest-coverage/ && npm run test:unit > jest-coverage/coverage.txt && npm run test:e2e"
```
@vickywane looking closer at the Cypress and Jest documentation, it appears like we should be uploading as github artifact a text or HTML version of the report produced by nyc merge, instead of trying to use separate txt files. See here: https://istanbul.js.org/docs/advanced/coverage-object-report/

Cypress seems to only be able to output json or a [brief txt summary](https://docs.cypress.io/guides/tooling/code-coverage#See-code-coverage-summary). The latter is not very useful for debugging.

vickywane replied:
Based on recent resolutions, we are uploading the coverage report generated by `CodeCov` as an artifact for each CI run.

This will lead to the removal of the scripts which manually combine the reports. I am awaiting your confirmation in this [comment](https://github.com/ambianic/ambianic-ui/pull/562#issuecomment-886605719) before removing the scripts.

ivelin commented on tests/unit/integrations/auth0-wrapper.spec.js:
```diff
@@ -0,0 +1,111 @@
+      getUser: jest.fn().mockReturnValue(null)
+    }
+
+    wrapper.vm.$auth.handleRedirectCallback()
```
@vickywane what is the test assertion here?

vickywane replied:
This test case has been adjusted [here](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/tests/unit/integrations/auth0-wrapper.spec.js).

ivelin commented on tests/unit/integrations/auth0-wrapper.spec.js:
```diff
@@ -0,0 +1,111 @@
+    wrapper.vm.$auth.logout()
+    wrapper.vm.$auth.authenticateUser()
+
+    done()
```
@vickywane what is the test assertion?

vickywane replied:
This test case has been adjusted [here](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/tests/unit/integrations/auth0-wrapper.spec.js).

ivelin commented on tests/unit/integrations/subscription-dialog.spec.js:
```diff
@@ -0,0 +1,112 @@
+
+    await wrapper.find('#confirm-btn').trigger('click')
+
+    fetch.mockReject(new Error('Error creating subscription'))
```
@vickywane what is the assertion for this test?

ivelin reviewed (changes requested):
@vickywane there seems to be still a lot of tests without proper assertion testing. Also coverage reporting and upload does not look right.

vickywane commented:
> @vickywane since this is the merged codecov report, I don't understand why you renamed it ./jest-coverage. That seems like a confusing and unnecessary change. 

The change was made while combining the jest and cypress reports. I have restored it back to the previous directory name.

vickywane commented:
> @vickywane looking closer at the Cypress and Jest documentation, it appears like we should be uploading as github artifact a text or HTML version of the report produced by nyc merge, instead of trying to use separate txt files. See here: https://istanbul.js.org/docs/advanced/coverage-object-report/

Alright. 

I would update the upload step within the workflow build to upload the merged result ( `out.json` )

vickywane commented:
> @vickywane should there be an assertion here to ensure the upgrade icon is displayed instead of the upgrade text?

The `[checkViewPort](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/cypress/integration/ambianic-tests/navbar.spec.js#L3)` function asserts that the upload icon ( from custom image ) is shown across the viewport passed in as a parameter. 

I extracted it into a function to avoid needless repetition.

vickywane commented:
> @vickywane the test should end with a meaningful assertion. Maybe check that the input field value is equal to the text that was entered and the submit button is now enabled after all required fields are filled with valid data.

I have updated the test to ensure that the typed in value is same as the value in the input. 
The equivalent unit test `subscription.spec.js` asserts that the value is typed-in is same as whats in local state.

vickywane commented:
> @vickywane what is the assertion for this test?

I have updated this test to assert that the subscription error text is shown whenever a network error occurs while creating a subscription.

vickywane commented:
@ivelin 

I have made changes based on your last review. 

As a rule of thumb, all new test suites now end with an assertion. Also, I have renamed most of the tests to have more descriptive names. 

The merged jest + cypress report is now being uploaded as an artifact on each CI build.

ivelin reviewed (changes requested):
@vickywane there are still a lot of unresolved conversations. Please take a look at each and provide your comment so we can resolve them. There are currently 23 unresolved conversations in this PR. Click on Conversations to see the list.
<img width="718" alt="Screen Shot 2021-07-09 at 3 33 44 AM" src="https://user-images.githubusercontent.com/2234901/125006406-add7a480-e023-11eb-81d1-3ae583dc8566.png">

ivelin commented on .github/workflows/main.yml:
```diff
@@ -43,6 +43,11 @@ jobs:
       env:
           CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    - name: Upload Jest coverage to Github Artifacts
```
@vickywane should this name value be updated to "Upload combined test coverage report ..." ?

vickywane replied:
Yes, i have renamed this job.

ivelin commented on package.json:
```diff
@@ -97,7 +103,7 @@
   "license": "Apache-2.0",
   "nyc": {
-    "report-dir": "coverage",
+    "report-dir": "cypress-coverage",
```
@vickywane what is the purpose of setting nyc's report-dir parameter to the cypress-coverage dir?

vickywane replied:
Initially i wanted to create seperate folders for the two reports. I have restored it back to the `coverage` name.

ivelin commented on reports/from-cypress.json:
```diff
@@ -0,0 +1,40 @@
+{"/home/victory/Desktop/OSS/ambianic-ui/cypress/integration/ambianic-tests/about.spec.js": {"path":"/home/victory/Desktop/OSS/ambianic-ui/cypress/integration/ambianic-tests/about.spec.js","statementMap":{"0":{"start":{"line":2,"column":0},"end":{"line":33,"column":2}},"1":{"start":{"line":3,"column":2},"end":{"line":5,"column":4}},"2":{"start":{"line":4,"column":4},"end":{"line":4,"column":22}}," ...
```
@vickywane what is the purpose of adding a coverage report to the github repo? Did you mean to add these files to .gitignore?

ivelin reviewed (changes requested):
@vickywane a few minor comments remain. Also the lighthouse errors needs to be fixed.

ivelin commented on .github/workflows/main.yml:
```diff
@@ -43,6 +43,11 @@ jobs:
       env:
           CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    - name: Upload Jest coverage to Github Artifacts
```
@vickywane did you see this comment?

vickywane replied:
Yes. 

I have corrected the steps. 
As discussed earlier, `CodeCov` is generating the coverage report and the generated file is being uploaded by the Upload artifact [here](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/.github/workflows/main.yml#L70).

ivelin commented on .github/workflows/main.yml:
```diff
@@ -60,10 +65,10 @@ jobs:
+          ./jest-coverage/coverage-final.json
         fail_ci_if_error: true # optional (default = false)
-        path_to_write_report: ./coverage/codecov_report.txt
+        path_to_write_report: ./jest-coverage/codecov_report.txt
```
@vickywane did you see this comment?

ivelin commented on .github/workflows/main.yml:
```diff
@@ -60,8 +65,8 @@ jobs:
+          ./cypress-coverage/coverage-final.json
+          ./jest-coverage/coverage-final.json
         fail_ci_if_error: true # optional (default = false)
         path_to_write_report: ./coverage/codecov_report.txt
```
@vickywane Looking at the codecov docs again, I think we can use this file to directly upload it as a combined coverage report artifact for github action and skip all the nyc merging code in this action workflow. Worth a try. Here is some example docs to look at:
https://github.com/codecov/codecov-action#example-workflowyml-with-codecov-action

ivelin replied:
@vickywane Did you see this comment from 15 days ago? 
If you looked at the `out.json` file currently uploaded as a github artifact, you will notice that it is not convenient to read. Let's see if we can use a txt or html format that allows us to quickly understand the coverage report.

ivelin commented on package.json:
```diff
@@ -7,14 +7,17 @@
-    "codecov:e2e": "codecov -f ./coverage/coverage-final.json",
-    "codecov:unit": "codecov -f ./coverage/jest/coverage-final.json",
-    "codecov": "codecov -f coverage/coverage-final.json && codecov -f coverage/jest/coverage-final.json",
+    "codecov:e2e": "codecov -f ./jest-coverage/coverage-final.json",
```
@vickywane did you see this comment ^^

ivelin replied:
@vickywane Reminder. This comment was made 25 days ago...

ivelin reviewed (changes requested):
Found a few more unaddressed and unresolved comments.

vickywane commented:
@ivelin I have created [this separate issue](https://github.com/ambianic/ambianic-ui/issues/671) to track the Lighthouse error. 

I will fix it today in a new pull request and revert to this.

vickywane commented:
> @vickywane what is the purpose of adding a coverage report to the GitHub repo? Did you mean to add these files to .gitignore?

The `reports` directory is already included in the `.gitignore` file.

vickywane commented:
> @vickywane should this name value be updated to "Upload combined test coverage report ..." ?

I have renamed the job.

vickywane commented:
> @vickywane what is the purpose of setting nyc's report-dir parameter to the cypress-coverage dir?

I wanted to split the two tests into separate directories.

ivelin commented:
@vickywane this PR is now out of sync with upstream. Please re-sync.

ivelin commented on package.json:
```diff
@@ -7,14 +7,17 @@
-    "codecov:unit": "codecov -f ./coverage/jest/coverage-final.json",
-    "codecov": "codecov -f coverage/coverage-final.json && codecov -f coverage/jest/coverage-final.json",
+    "codecov:e2e": "codecov -f ./jest-coverage/coverage-final.json",
+    "codecov:unit": "codecov -f ./jest-coverage/coverage-final.json",
```
@vickywane `codecov:unit` and `codecov` appear to be stale. Are these used anywhere?

vickywane replied:
No. They are not being used by the application itself, except by other contributors.

I have removed the `codecov:unit` script.

ivelin reviewed (changes requested):
@vickywane see open comments.

vickywane commented:
> `codecov:unit` and `codecov` appear to be stale. Are these used anywhere?

No they are not being used. Alongside with `codecov:e2e`. 

I will proceed to remove them since they are not being used.

vickywane commented on cypress/integration/ambianic-tests/authbar-menu.spec.js:
```diff
@@ -0,0 +1,53 @@
+
+    profileBtn.should('be.visible')
+    profileBtn.children('.user-img').should('have.length', 1)
+    profileBtn.click()
```
This has been corrected. 

I refactored all tests written to end with different [assertions](https://github.com/vickywane/ambianic-ui/blob/feature/user-management/cypress/integration/ambianic-tests/authbar-menu.spec.js).
